### ISSUE TITLE
x/build: add LUCI aix-ppc64 builder

### ISSUE THREAD

**Issue by dmitshur** [golang/go#67299]
There currently isn't a LUCI builder that tests the aix/ppc64 port (other than the misc-compile builder, which tests only that the port compiles). This is the tracking issue for it.

The next steps that a builder owner will need to follow to make progress here are documented https://go.dev/wiki/DashboardBuilders#luci-builders.

**Comment by laboger** [comment]
@ayappanec

**Comment by ayappanec** [comment]
We are working on getting a AIX machine in OSU lab for this purpose.

**Comment by dmitshur** [comment]
CC @golang/aix.

**Comment by pmur** [comment]
I'll try poking around with this since I came into access to the VM.

I've chosen the hostname as `aix-ppc64-power8`, cert attached: [aix-ppc64-power8.csr.txt](https://github.com/user-attachments/files/16367309/aix-ppc64-power8.csr.txt)

**Comment by dmitshur** [comment]
@pmur I see you've edited that post, replacing `-` with `_`. You were likely thinking of builder names, where underscores are indeed used to separate the OS version and/or runmods. But for hostnames we do use dashes everywhere. The ideal format is documented at https://go.dev/wiki/DashboardBuilders#how-to-set-up-a-builder as `<GOOS>-<GOOARCH>-<GitHub handle of maintainer>`. If it's not definitive who the maintainer will be, it's fine to go with just `aix-ppc64-power8` as before since this is the first aix/ppc64 LUCI builder.

Are you okay to go back to `aix-ppc64-power8` (or `aix-ppc64-power8-suffix`) as the hostname?

**Comment by pmur** [comment]
Oops. I'd prefer to stick with `aix-ppc64-power8` for consistency with the other builders I've inherited.

**Comment by dmitshur** [comment]
Thanks. Here's the certificate: [aix-ppc64-power8-1721862086.cert.txt](https://github.com/user-attachments/files/16368164/aix-ppc64-power8-1721862086.cert.txt).

**Comment by pmur** [comment]
I am still in the process of getting LUCI working. I had to update the OSS software, and that has been rather bumpy. I have the swarming client running, though I suspect I will need to solve more issues once the backend pieces are updated.

This has led to some (hopefully) [intermittent failures](https://build.golang.org/log/85fccbf73cead95e05a32d7e7eafc08912ccd88e) of the old builder.

**Comment by pmur** [comment]
The LUCI aix bot is [up](https://chromium-swarm.appspot.com/bot?id=aix-ppc64-power8), though I think something needs to be updated on the backend. It has not run any jobs yet. The cpu and cipd_platform values look suspect.

**Comment by dmitshur** [comment]
> The cpu and cipd_platform values look suspect.

Indeed. The CIPD binaries built for GOOS=aix GOARCH=ppc64 are using "aix-ppc64" (https://chrome-infra-packages.appspot.com/p/infra/experimental/golangbuild/aix-ppc64), so it's intended for the bot to report its cipd_platform dimension as that (instead of the current "aix-00fac25f4b00"). That dimension is also used for match the bot to the builder definitions (e.g., [gotip-aix-ppc64](https://ci.chromium.org/ui/p/golang/builders/ci/gotip-aix-ppc64). There are zero matches at https://chromium-swarm.appspot.com/botlist?f=cipd_platform%3Aaix-ppc64&f=pool%3Aluci.golang.shared-workers now—hence the bot isn't getting any jobs.

**Comment by dmitshur** [comment]
The aforementioned issue is fixed in [crrev.com/c/5775509](https://crrev.com/c/5775509), thanks. A build for gotip-aix-ppc64 builder was now able to start at https://ci.chromium.org/b/8739762966790400417, failing with "CIPD error: invalid_version_error with package infra/tools/luci/cas/aix-ppc64, version git_revision:2aba496613f92a5b06d577f82b5d028225d3d577 on path None".

While we do have a aix-ppc64 binary for golangbuild (https://chrome-infra-packages.appspot.com/p/infra/experimental/golangbuild), there isn't one for some other dependencies like `cas` at https://chrome-infra-packages.appspot.com/p/infra/tools/luci/cas. We either need to get them built and selected, or possibly get by if there's a way this builder can operate in "low dependency" mode without needing them.

**Comment by dmitshur** [comment]
Other binaries like [bb](https://chrome-infra-packages.appspot.com/p/infra/tools/bb/aix-ppc64), [rdb](https://chrome-infra-packages.appspot.com/p/infra/tools/rdb/aix-ppc64), [result_adapter](https://chrome-infra-packages.appspot.com/p/infra/tools/result_adapter/aix-ppc64), and so on are otherwise there. The reason we didn't end up with a cas binary built for aix/ppc64 is because it fails to build:

```
$ GOOS=aix GOARCH=ppc64 CGO_ENABLED=0 go install go.chromium.org/luci/client/cmd/cas@latest
# github.com/dgraph-io/badger/v3
go/pkg/mod/github.com/dgraph-io/badger/v3@v3.2103.2/dir_unix.go:62:13: undefined: unix.Flock
```

**Comment by pmur** [comment]
@dmitshur are there any other dependencies which do not compile for aix?

I have an attempt to get badger (v4) running on aix, but I need to get approval to post and contribute the patch. I am not sure how long that will take.

**Comment by dmitshur** [comment]
Thanks for working on this Paul. From what I can tell this case (`cas` failing to build for aix/ppc64) is the only known outstanding issue for this port. It's a good idea to check ahead for possible future problems, and I'm not aware of something more at this time.

**Comment by aclements** [comment]
FYI, the aix/ppc64 is now passing on the dashboard.

**Comment by dmitshur** [comment]
@aclements You're referring to the legacy aix/ppc64 builder, right? This issue tracks the new LUCI builder which isn't [working yet](https://ci.chromium.org/ui/p/golang/builders/luci.golang.ci/gotip-aix-ppc64). It looks like there has been progress since September 4 in that https://github.com/dgraph-io/badger/pull/2115 exists to fix the build in a `cas` dependency, but it hasn't been reviewed yet.

**Comment by aclements** [comment]
@dmitshur , yes, you're right. I was looking at the legacy dashboard, which I see only shows the legacy aix/ppc64 builder.

**Comment by dmitshur** [comment]
As I understand that PR applied the fix to the v4 major version of badger, but cas currently uses v3. That means either the badger fix needs to also be backported to its v3, and cas updated to that, or otherwise cas would need to move up to use v4. It's probably worth taking a look at both of those possible paths forward and see which one is better to pursue.

**Comment by ayappanec** [comment]
> [@ayappanec](https://github.com/ayappanec) could you verify the luci CI components are still set up and running on the vm? I no longer have permissions to inspect the luci webui.

Sure. Let me check

**Comment by ayappanec** [comment]
@pmur In the aix vm ( where the legacy builder is running ), I see keys and luci_machine_tokend binary available. But I don't see any process related to luci running.

**Comment by pmur** [comment]
@ayappanec I may have tried to get it running year(s?) ago, but never got beyond the issue with badger.

The setup instructions should loosely follow those at https://go.dev/wiki/DashboardBuilders. However the binary mentioned early needs fixed. I am unable to help fix that binary for aix.

**Comment by ayappanec** [comment]
Thanks @pmur. I understand

@dmitshur 
I tried starting boostrapswarm and looking at the swarming logs , I see the below 

```
11469174 2026-04-16 12:05:00.282 I: Subprocess for task_runner started.
11469174 2026-04-16 12:05:09.631 I: task_runner exit: 0
11469174 2026-04-16 12:05:09.633 I: Calling /swarming/api/v1/bot/rbe/session/update
11469174 2026-04-16 12:05:09.634 D: GOOGAPPUID = sha1(2026-04-16-aix-ppc64-power8:/home/swarming/.swarming) % 1000 = 450
11469174 2026-04-16 12:05:09.768 D: https://chromium-swarm.appspot.com:443 "POST /swarming/api/v1/bot/rbe/session/update HTTP/1
.1" 200 None
11469174 2026-04-16 12:05:09.769 D: Request https://chromium-swarm.appspot.com/swarming/api/v1/bot/rbe/session/update succeeded
11469174 2026-04-16 12:05:09.769 D: RBE projects/chromium-swarm/instances/default_instance/botSessions/3989470878035416396: RBE
SessionStatus.MAINTENANCE => RBESessionStatus.OK
11469174 2026-04-16 12:05:09.770 I: calling on_after_task: failure=False, internal_failure=True, task_dimensions={'cipd_platfor
m': ['aix-ppc64'], 'pool': ['luci.golang.shared-workers']}, task_result={'exit_code': None, 'hard_timeout': False, 'io_timeout'
: False, 'internal_error_reported': True, 'internal_error': 'CIPD error: invalid_version_error with package infra/tools/luci/bb
agent/aix-ppc64, version git_revision:4e276d2c082f0b35a37e93d86b9cd78c7eb45f6e on path None', 'version': 3}
11469174 2026-04-16 12:05:09.770 I: [mmutex] on_after_task releasing mutex
11469174 2026-04-16 12:05:09.770 I: [mmutex] Not releasing maintenance mutex; not held
11469174 2026-04-16 12:05:09.774 I: [mmutex] Reboot message: Internal failure
11469174 2026-04-16 12:05:09.774 I: [mmutex] Restart message: None
11469174 2026-04-16 12:05:09.774 I: RBE: terminating session projects/chromium-swarm/instances/default_instance/botSessions/398
9470878035416396 with status RBESessionStatus.HOST_REBOOTING
11469174 2026-04-16 12:05:09.775 I: Calling /swarming/api/v1/bot/rbe/session/update
11469174 2026-04-16 12:05:09.775 D: GOOGAPPUID = sha1(2026-04-16-aix-ppc64-power8:/home/swarming/.swarming) % 1000 = 450
11469174 2026-04-16 12:05:09.902 D: https://chromium-swarm.appspot.com:443 "POST /swarming/api/v1/bot/rbe/session/update HTTP/1
.1" 200 None
11469174 2026-04-16 12:05:09.903 D: Request https://chromium-swarm.appspot.com/swarming/api/v1/bot/rbe/session/update succeeded
11469174 2026-04-16 12:05:09.903 D: RBE projects/chromium-swarm/instances/default_instance/botSessions/3989470878035416396: RBE
SessionStatus.HOST_REBOOTING => RBESessionStatus.OK
11469174 2026-04-16 12:05:09.904 I: Unknown OS aix
11469174 2026-04-16 12:05:09.905 I: on_bot_shutdown(): 0.001s
```

It shows "CIPD error: invalid_version_error with package infra/tools/luci/bbagent/aix-ppc64" .
Is it that bbagent binary is also not available for aix-ppc64 ?

**Comment by dmitshur** [comment]
That message does seem unexpected. I also see it in a recent task like https://chromium-swarm.appspot.com/task?id=77aef1f18b21ad10 and it's blocking further progress on the task.

If you look at https://chrome-infra-packages.appspot.com/p/infra/tools/luci/bbagent/aix-ppc64 you should see there are current versions of that package available, matching that of e.g. https://chrome-infra-packages.appspot.com/p/infra/tools/luci/bbagent/linux-amd64, but not the particular `git_revision:4e276d2c082f0b35a37e93d86b9cd78c7eb45f6e` version. I'll look into this.

Noting that in addition to this `bbagent` issue needing to be resolved now, the `cas` package mentioned in https://github.com/golang/go/issues/67299#issuecomment-3849808228 still likely needs attention.

**Comment by ayappanec** [comment]
Thanks @dmitshur 
For the `cas` issue, you mentioned two approaches "either the badger fix needs to also be backported to its v3, and cas updated to that, or otherwise cas would need to move up to use v4" 
Backporting the fix to v3 also requires `badger` community releasing a new v3 version ? I am not sure whether that will happen. The last v3 version released some 4 years back. 
Regarding the other approach, I updated `badger` to the latest v4.9.1 in `cas` and ran the inbuilt test suites. The results are all passing.

In /luci-go/client/cmd/cas/casimpl,
```
# go test -v 
=== RUN   TestGetRoot
=== PAUSE TestGetRoot
=== RUN   TestLoadPathsJSON
=== PAUSE TestLoadPathsJSON
=== RUN   TestArchiveDownload
=== PAUSE TestArchiveDownload
=== RUN   TestScatterGatherAdd
=== PAUSE TestScatterGatherAdd
=== RUN   TestScatterGatherSet
=== PAUSE TestScatterGatherSet
=== CONT  TestGetRoot
=== RUN   TestGetRoot/Basic
--- PASS: TestGetRoot (0.00s)
    --- PASS: TestGetRoot/Basic (0.00s)
=== CONT  TestScatterGatherAdd
=== RUN   TestScatterGatherAdd/Test_that_Add_works_in_a_good_case.
=== RUN   TestScatterGatherAdd/Test_that_Add_fails_in_a_bad_case.
--- PASS: TestScatterGatherAdd (0.00s)
    --- PASS: TestScatterGatherAdd/Test_that_Add_works_in_a_good_case. (0.00s)
    --- PASS: TestScatterGatherAdd/Test_that_Add_fails_in_a_bad_case. (0.00s)
=== CONT  TestArchiveDownload
=== RUN   TestArchiveDownload/Upload_and_download
=== CONT  TestLoadPathsJSON
=== RUN   TestLoadPathsJSON/Basic
=== CONT  TestScatterGatherSet
=== RUN   TestScatterGatherSet/Test_that_Set_works_in_a_good_case.
--- PASS: TestScatterGatherSet (0.00s)
    --- PASS: TestScatterGatherSet/Test_that_Set_works_in_a_good_case. (0.00s)
--- PASS: TestLoadPathsJSON (0.26s)
    --- PASS: TestLoadPathsJSON/Basic (0.26s)
=== RUN   TestArchiveDownload/Upload_and_download/use_cache
=== RUN   TestArchiveDownload/Upload_and_download/not_use_cache
--- PASS: TestArchiveDownload (2.51s)
    --- PASS: TestArchiveDownload/Upload_and_download (2.51s)
        --- PASS: TestArchiveDownload/Upload_and_download/use_cache (1.54s)
        --- PASS: TestArchiveDownload/Upload_and_download/not_use_cache (0.36s)
PASS
ok      go.chromium.org/luci/client/cmd/cas/casimpl     2.650s
```

In /luci-go/client/casclient.,
```
# go test -v 
=== RUN   TestIsLocalAddr
=== PAUSE TestIsLocalAddr
=== RUN   TestParseCASInstance
=== PAUSE TestParseCASInstance
=== CONT  TestIsLocalAddr
=== RUN   TestIsLocalAddr/isLocalAddr
=== CONT  TestParseCASInstance
=== RUN   TestParseCASInstance/Basic
=== RUN   TestParseCASInstance/Invalid
--- PASS: TestParseCASInstance (0.00s)
    --- PASS: TestParseCASInstance/Basic (0.00s)
    --- PASS: TestParseCASInstance/Invalid (0.00s)
--- PASS: TestIsLocalAddr (0.05s)
    --- PASS: TestIsLocalAddr/isLocalAddr (0.05s)
PASS
ok      go.chromium.org/luci/client/casclient   0.138s
```
In /luci-go/common/data/embeddedkvs.,
```
# go test -v 
=== RUN   TestCache
=== PAUSE TestCache
=== CONT  TestCache
=== RUN   TestCache/basic
--- PASS: TestCache (0.31s)
    --- PASS: TestCache/basic (0.31s)
PASS
ok      go.chromium.org/luci/common/data/embeddedkvs    0.357s

```
